pdg/src/info.rs

```diff
@@ -52,152 +183,413 @@ fn node_does_neg_offset(n: &Node) -> bool {
-/// the correct flow information.
-pub fn augment_with_info(pdg: &mut Graphs) {
-    let gs: &mut IndexVec<GraphId, Graph> = &mut pdg.graphs;
-    for g in gs.iter_mut() {
-        let mut idx_flow_to_mut = HashMap::new();
-        let mut idx_flow_to_use = HashMap::new();
-        let mut idx_flow_to_pos_offset = HashMap::new();
-        let mut idx_flow_to_neg_offset = HashMap::new();
-        let mut idx_non_unique = HashMap::new();
-        for (idx, _) in g.nodes.iter_enumerated() {
-            if let Some(descmutidx) = check_flows_to_node_kind(g, &idx, node_does_mutation) {
-                idx_flow_to_mut.insert(idx, descmutidx);
-            }
-            if let Some(descuseidx) = check_flows_to_node_kind(g, &idx, node_does_load) {
-                idx_flow_to_use.insert(idx, descuseidx);
-            }
-            if let Some(descposoidx) = check_flows_to_node_kind(g, &idx, node_does_pos_offset) {
-                idx_flow_to_pos_offset.insert(idx, descposoidx);
-            }
-            if let Some(descnegoidx) = check_flows_to_node_kind(g, &idx, node_does_neg_offset) {
-                idx_flow_to_neg_offset.insert(idx, descnegoidx);
-            }
-            if let Some(non_unique_idx) = check_whether_rules_obeyed(g, &idx) {
-                idx_non_unique.insert(idx, non_unique_idx);
-            }
-        }
-        for (idx, node) in g.nodes.iter_enumerated_mut() {
-            node.node_info = Some(NodeInfo {
-                flows_to_mutation: idx_flow_to_mut.remove(&idx),
-                flows_to_load: idx_flow_to_use.remove(&idx),
-                flows_to_pos_offset: idx_flow_to_pos_offset.remove(&idx),
-                flows_to_neg_offset: idx_flow_to_pos_offset.remove(&idx),
-                non_unique: idx_non_unique.remove(&idx),
-            })
-        }
+    fn mk_load_addr(g: &mut Graph, source: NodeId) -> NodeId {
+        mk_node(g, NodeKind::LoadAddr, Some(source))
+    }
+
+    fn mk_store_addr(g: &mut Graph, source: NodeId) -> NodeId {
+        mk_node(g, NodeKind::StoreAddr, Some(source))
+    }
+
+    fn mk_field(g: &mut Graph, source: NodeId, field: impl Into<Field>) -> NodeId {
+        mk_node(g, NodeKind::Field(field.into()), Some(source))
+    }
+
+    fn build_pdg(g: Graph) -> Graphs {
+        let mut pdg = Graphs::default();
+        pdg.graphs.push(g);
+        augment_with_info(&mut pdg);
+        pdg
+    }
+
+    fn info(pdg: &Graphs, id: NodeId) -> &NodeInfo {
+        pdg.graphs[0_u32.into()].nodes[id]
+            .node_info
+            .as_ref()
+            .unwrap()
+    }
+
+    #[test]
+    fn unique_interleave() {
+        let mut g = Graph::default();
+        // let mut a = 0;   // A
+        // let b = &mut a;  // B1
+        // *b = 0;          // B2
+        // let c = &mut a;  // C1
+        // *c = 0;          // C2
+        // *b = 0;          // B3
+        // *c = 0;          // C3
```

Comment:
You could put this in a doc comment for the function and have it run as a doc test.  Same for all of these test functions.